Crystal interpreter undefined reference to `GC_malloc_atomic'

my fork of the Crystal compiler has the interpreter enabled but it doesn't function correctly on Windows
```
$ crystal i
Crystal interpreter 1.10.1 [c6f3552f5] (2023-10-13).
EXPERIMENTAL SOFTWARE: if you find a bug, please consider opening an issue in
https://github.com/crystal-lang/crystal/issues/new/
undefined reference to `GC_malloc_atomic' (Crystal::Loader::LoadError)
  from D:\a\crystal\crystal\src\compiler\crystal\loader.cr:77 in 'find_symbol'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:1861 in 'visit'
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:244 in 'compile_def'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:2079 in 'create_compiled_def'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:1874 in 'visit'
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:244 in 'compile_def'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:2079 in 'create_compiled_def'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:1874 in 'visit'
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:244 in 'compile_def'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:2079 in 'create_compiled_def'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:1874 in 'visit'
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:244 in 'compile_def'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:2079 in 'create_compiled_def'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:1874 in 'visit'
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:2004 in 'create_compiled_def'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:1874 in 'visit'
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:244 in 'compile_def'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:2079 in 'create_compiled_def'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:1874 in 'visit'
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:244 in 'compile_def'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:2079 in 'create_compiled_def'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:1874 in 'visit'
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:244 in 'compile_def'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:1473 in 'get_const_index_and_compiled_def'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:1445 in 'visit'
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:1870 in 'visit'
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:244 in 'compile_def'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:1473 in 'get_const_index_and_compiled_def'
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:3126 in 'visit'
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:3126 in 'visit'
  from D:\a\crystal\crystal\src\compiler\crystal\syntax\visitor.cr:27 in 'accept'     
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\compiler.cr:161 in 'compile'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\interpreter.cr:232 in 'interpret'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\repl.cr:98 in 'interpret_and_exit_on_error'
  from D:\a\crystal\crystal\src\compiler\crystal\interpreter\repl.cr:17 in 'run'      
  from D:\a\crystal\crystal\src\compiler\crystal\command.cr:103 in 'run'
  from D:\a\crystal\crystal\src\compiler\crystal.cr:11 in '__crystal_main'
  from D:\a\crystal\crystal\src\crystal\main.cr:141 in 'main'
  from D:\a\crystal\crystal\src\crystal\system\win32\wmain.cr:37 in 'wmain'
  from D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288 in '__scrt_common_main_seh'
  from C:\WINDOWS\System32\KERNEL32.DLL +95748 in 'BaseThreadInitThunk'
  from C:\WINDOWS\SYSTEM32\ntdll.dll +337569 in 'RtlUserThreadStart'
```
link to fork and binaries used:
https://github.com/Epikest/crystal
https://github.com/Epikest/crystal/releases/tag/1.10.1
```
$ crystal version
Crystal 1.10.1 [c6f3552f5] (2023-10-13)

LLVM: 17.0.2
Default target: x86_64-pc-windows-msvc
```

--- a/spec/compiler/loader/msvc_spec.cr
+++ b/spec/compiler/loader/msvc_spec.cr
@@ -125,4 +125,30 @@ describe Crystal::Loader do
       end
     end
   end
+
+  describe "lib suffix" do
+    before_all do
+      FileUtils.mkdir_p(SPEC_CRYSTAL_LOADER_LIB_PATH)
+    end
+
+    after_all do
+      FileUtils.rm_rf(SPEC_CRYSTAL_LOADER_LIB_PATH)
+    end
+
+    it "respects -dynamic" do
+      build_c_dynlib(compiler_datapath("loader", "foo.c"), lib_name: "foo-dynamic")
+      loader = Crystal::Loader.new([SPEC_CRYSTAL_LOADER_LIB_PATH] of String)
+      loader.load_library?("foo").should be_true
+    ensure
+      loader.close_all if loader
+    end
+
+    it "ignores -static" do
+      build_c_dynlib(compiler_datapath("loader", "foo.c"), lib_name: "bar-static")
+      loader = Crystal::Loader.new([SPEC_CRYSTAL_LOADER_LIB_PATH] of String)
+      loader.load_library?("bar").should be_false
+    ensure
+      loader.close_all if loader
+    end
+  end
 end

--- a/spec/compiler/loader/spec_helper.cr
+++ b/spec/compiler/loader/spec_helper.cr
@@ -2,8 +2,8 @@ require "spec"
 
 SPEC_CRYSTAL_LOADER_LIB_PATH = File.join(SPEC_TEMPFILE_PATH, "loader")
 
-def build_c_dynlib(c_filename, target_dir = SPEC_CRYSTAL_LOADER_LIB_PATH)
-  o_filename = File.join(target_dir, Crystal::Loader.library_filename(File.basename(c_filename, ".c")))
+def build_c_dynlib(c_filename, *, lib_name = nil, target_dir = SPEC_CRYSTAL_LOADER_LIB_PATH)
+  o_filename = File.join(target_dir, Crystal::Loader.library_filename(lib_name || File.basename(c_filename, ".c")))
 
   {% if flag?(:msvc) %}
     o_basename = o_filename.rchop(".lib")

--- a/src/compiler/crystal/loader/msvc.cr
+++ b/src/compiler/crystal/loader/msvc.cr
@@ -60,7 +60,7 @@ class Crystal::Loader
     result
   end
 
-  private def self.search_library(libname, search_paths, extra_suffix)
+  protected def self.search_library(libname, search_paths, extra_suffix)
     if ::Path::SEPARATORS.any? { |separator| libname.includes?(separator) }
       libname = File.expand_path(libname)
       library_path = library_filename(libname)
@@ -106,7 +106,7 @@ class Crystal::Loader
   end
 
   def self.library_filename(libname : String) : String
-    "#{libname}.lib"
+    "#{libname.rchop(".lib")}.lib"
   end
 
   def find_symbol?(name : String) : Handle?
@@ -149,8 +149,13 @@ class Crystal::Loader
     load_library?(libname) || raise LoadError.from_winerror "cannot find #{Loader.library_filename(libname)}"
   end
 
+  def load_library?(libname : String) : Bool
+    library_path = Loader.search_library(libname, @search_paths, "-dynamic")
+    !library_path.nil? && load_file?(library_path)
+  end
+
   private def open_library(path : String)
-    # TODO: respect Crystal::LIBRARY_RPATH (#13490)
+    # TODO: respect Crystal::LIBRARY_RPATH (#13490), or `@[Link(dll:)]`'s search order
     LibC.LoadLibraryExW(System.to_wstr(path), nil, 0)
   end
 

